### PR TITLE
Add DoS resistance for v0.11

### PR DESCRIPTION
**Pull request by terencechain**
Part of #5119 

And part of ethereum/eth2.0-specs#1615

This PR updated sync with the following dos preventions:
- multiple beacon blocks from same proposer in a slot
- multiple attestations from same participant in a slot
- multiple aggregate_and_proofs from same aggregator in a slot

### CODE REVIEW

File: beacon-chain/sync/validate_aggregate_proof.go [hunk 1]
```diff
@@ -44,32 +44,39 @@ func (r *Service) validateAggregateAndProof(ctx context.Context, pid peer.ID, ms
...
 		return false
 	}
-	m, ok := raw.(*ethpb.AggregateAttestationAndProof)
+	m, ok := raw.(*ethpb.SignedAggregateAttestationAndProof)
 	if !ok {
 		return false
 	}
 
+	// Verify this is the first aggregate received from the aggregator with index and slot.
+	if r.hasSeenAggregatorIndexSlot(m.Message.Aggregate.Data.Slot, m.Message.AggregatorIndex) {
```
Comment (prestonvanloon) [inline beacon-chain/sync/validate_aggregate_proof.go]: Can you check these are set? Looks like a prime target for nil pointer panic

File: beacon-chain/sync/validate_aggregate_proof.go [hunk 2]
```diff
@@ -44,32 +44,39 @@ func (r *Service) validateAggregateAndProof(ctx context.Context, pid peer.ID, ms
...
+	if !r.validateAggregatedAtt(ctx, m.Message) {
 		return false
 	}
 
-	if !featureconfig.Get().DisableStrictAttestationPubsubVerification && !r.chain.IsValidAttestation(ctx, m.Aggregate) {
+	if !featureconfig.Get().DisableStrictAttestationPubsubVerification && !r.chain.IsValidAttestation(ctx, m.Message.Aggregate) {
 		return false
 	}
 
+	r.setAggregatorIndexSlotSeen(m.Message.Aggregate.Data.Slot, m.Message.AggregatorIndex)
```
Comment (prestonvanloon) [inline beacon-chain/sync/validate_aggregate_proof.go]: Shouldn't this be done in the subscriber?

File: beacon-chain/sync/service.go [hunk 3]
```diff
@@ -100,10 +103,33 @@ type Service struct {
...
+		NumCounters: seenBlockSize, // Max size is seenBlockSize * 32byte.
+		MaxCost:     seenBlockSize / 10,
+		BufferItems: 64,
+	})
+	if err != nil {
+		panic(err)
+	}
+	aCache, err := ristretto.NewCache(&ristretto.Config{
+		NumCounters: seenAttSize, // Max size is seenAttSize * 32byte.
+		MaxCost:     seenAttSize / 10,
```
Comment (prestonvanloon) [inline beacon-chain/sync/service.go]: You are costing values as 1. What is the intended number of values to keep in the cache?

Comment (terencechain) [reply beacon-chain/sync/service.go]: I'm budgeting 100 for blocks and 1000 for attestations. With proper dynamic committee subnet, there's roughly 10x more attestations than block on per slot level coming in via the wire. I definitely over-budged a bit but these are cheap, just a 32byte key to bool

File: beacon-chain/sync/validate_aggregate_proof.go (hunk 1 above)
Comment (terencechain) [reply beacon-chain/sync/validate_aggregate_proof.go]: sounds good

File: beacon-chain/sync/validate_aggregate_proof.go (hunk 2 above)
Comment (terencechain) [reply beacon-chain/sync/validate_aggregate_proof.go]: agree

File: beacon-chain/sync/service.go [hunk 4]
```diff
@@ -100,10 +103,33 @@ type Service struct {
...
 	attestationNotifier  operation.Notifier
+	seenBlockLock        sync.RWMutex
+	seenBlockCache       *ristretto.Cache
+	seenAttestationLock  sync.RWMutex
+	seenAttestationCache *ristretto.Cache
 }
 
 // Start the regular sync service.
 func (r *Service) Start() {
+	bCache, err := ristretto.NewCache(&ristretto.Config{
```
Comment (nisdas) [inline beacon-chain/sync/service.go]: I think another cache might have to be used for this, in ristretto `Sets` aren't guaranteed and for this usecase, it seems like we would want setting a value to be included in the cache

Comment (terencechain) [reply beacon-chain/sync/service.go]: Good point, switched to LRU. Please take a look

**Pull request merged**